`GEOMETRY` Rework: Part 6 - Geometry "Shredding"

Maxxen opened the pull request:
This is a followup PR that builds on top of https://github.com/duckdb/duckdb/pull/19848 (although orthogonal to https://github.com/duckdb/duckdb/pull/20143). Please have a look at https://github.com/duckdb/duckdb/pull/19136 for the context behind this PR.

This PR enables support for "shredding" geometry columns in our storage format by internally decomposing the geometries into multiple separate column segments when all the rows within a row-group are of the same geometry sub-type. For example, if a row group only contains `POINT` (XY) geometries, the column segment for that row-group gets rewritten and stored internally as `STRUCT(x DOUBLE, y DOUBLE)` instead of `BLOB` when checkpointing. This encoding is similar to how [GeoArrow]() encodes typed geometries.

The major benefit of "decomposing" or "shredding" geometries from blobs into columnar lists and structs of doubles like this is at the storage-level is that they compress _significantly_ better in the decomposed form. By decomposing the geometries into column segment of primitive types we automatically benefit from DuckDB's existing (and future!) best-in-class adaptive compression algorithms for each type of component. E.g. the coordinate vertices get ALP-compressed, and polygon ring-offsets get RLE/delta/dictionary encoded automatically based on the distribution of the data. 

In my (limited) testing, a fully shredded column is about half the size as the blob-equivalent, i.e. shredding results in a 2x compression ratio. This means that you get much faster reads from disk, and will be able to keep more data in memory (as we now also compress in-memory segments). 

There are currently two caveats: 
- We don't shred segments that contain `GEOMETRYCOLLECTION`s, as they can be recursive and don't have a fixed layout. 
- We only shred if the _entire_ row-group is of a single geometry sub-type. So if you have 10000 points, insert a linestring and checkpoint, the column segment will be reverted back to physical type `BLOB`. 
- We don't shred segments containing EMPTY geometries, either at the root or in sub-geometries.

We may want to consider "partial" shredding in the future, where we keep multiple shredded segments around. But Im not sure if the performance/complexity hit would be worth it. 

Despite those limitations, geometry shredding still opens up some interesting future optimizations opportunities:

- Push down certain filters (like bounding-box-intersections) much deeper into the storage as they are faster to evaluate on separate coordinate-arrays axis-by-axis.
- Emit "shredded" geometries directly from storage when casting or exporting to GeoArrow
  - Or if the optimizer realizes that all function expressions that read from a fully shredded column have overloads that take the shredded representation instead (e.g. spatial's `POINT_2D`), replace all functions (or insert casts where needed) and emit the shredded representation from storage to "specialize" queries automatically. 

As it stands, In the case where there is no shredding, there is no additional storage overhead. In other words, a non-shredded geometry column is serialized exactly the same way as it used to (before this PR lands). 

I've made some changes to how `ColumnData` is serialized though. There is now an abstract `unique_ptr<ExtraPersistentColumnData>` in the `PersistentColumnData` struct that separates out the extra info required by the `VariantColumnData` and `GeoColumnData`. As they are the only two column data types where the type/layout can differ between segments within the same column, they need to store some information to reconstruct the layout and can't just rely on the top-level column logical type. In the case of the variant, this is the "shredded" type, but since geometries have a fixed number of layouts, we only store the geometry and vertex-sub type (enums) instead of the equivalent logical type to save space. I expect that we may want to generalize this further in the future if we start implementing shredding to other types as well (e.g. strings/json).

Maxxen committed:
initial attempt at geometry POINT shredding

Maxxen committed:
fix polygon shredding

Maxxen committed:
rework geo column shredding entirely and simplify

Maxxen committed:
fix misc bugs with validity

Maxxen committed:
remove SplitColumnData. Fix serialization - but is ugly, needs to change

Maxxen committed:
add include

Maxxen committed:
add more shredding

Maxxen committed:
serialize shredding state, add support for shredding other geometry types too

Maxxen committed:
fixup variant

Maxxen committed:
shred multi-types too

Maxxen committed:
fix failing tests

Maxxen committed:
more fixes, hardcode variant for now

Maxxen committed:
require stats to work

Maxxen committed:
fix warnings/rebase changes

Maxxen committed:
deserialize stats correctly, add tests for all subtypes

Maxxen committed:
Add support for tracking empty geometry in stats - dont shred if there are empty points

Maxxen committed:
add empty_geom and empty_part to geometry stats and use it to avoid shredding empty geoms

Maxxen committed:
fix block verification

Maxxen committed:
add setting to control shredding threshold

Maxxen committed:
vector size

Maxxen committed:
dont checkpoint on shutdown in test

Mytherin reviewed (commented):
Thanks for the PR! Looks good - some comments

Maxxen committed:
push down stats when checkpointing

Maxxen committed:
Fix Fetch and FetchRow, add support to reassemble with offset

Maxxen committed:
change serialization id

Maxxen committed:
switch to generic setting

Maxxen committed:
rebase on top of v1.5

Maxxen commented:
@Mytherin I've adapted to your feedback, implemented the remaining borked methods and rebased to v1.5 - I'm gonna let it simmer on my own CI for a bit but let me know if there is anything else you think should be done.

Mytherin commented on src/storage/table/geo_column_data.cpp:
```diff
@@ -0,0 +1,552 @@
+	return scan_count;
+}
+//
+// idx_t GeoColumnData::ScanCommitted(idx_t vector_index, ColumnScanState &state, Vector &result, bool allow_updates,
```
nit: this can go

Mytherin commented on src/include/duckdb/storage/statistics/geometry_stats.hpp:
```diff
@@ -152,10 +291,13 @@ struct GeometryStats {
 	static GeometryStatsData &GetDataUnsafe(BaseStatistics &stats);
 	static const GeometryStatsData &GetDataUnsafe(const BaseStatistics &stats);
+
+private:
```
Should this be made public? If so - maybe just remove the empty private:?

Maxxen committed:
remove dead code

Maxxen committed:
fixup test

Maxxen committed:
Merge remote-tracking branch 'upstream/v1.5-variegata' into core-geom-step-5-shredding

Mytherin reviewed (commented):
Thanks! Looks good - some minor comments otherwise this is good to go

Maxxen committed:
feedback

Maxxen committed:
Merge branch 'v1.5-variegata' of https://github.com/duckdb/duckdb into core-geom-step-5-shredding
